Removing quantization from the dispatcher. Changing the message.

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29054 Removing quantization from the dispatcher. Changing the message.**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/29054

Differential Revision: [D18276881](https://our.internmc.facebook.com/intern/diff/D18276881)

Commit:
Removing quantization from the dispatcher. Changing the message.

z-a-f commented:
Fixes #28518

Commit:
Changed the tests on "Removing quantization from the dispatcher. Changing the message."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29054

Differential Revision: [D18276881](https://our.internmc.facebook.com/intern/diff/D18276881)

jamesr66a commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -291,25 +291,12 @@ class DispatchTable final {
+                  " registered as '", dispatch_key_str, "'. '", operator_name_,
+                  "' is registered with these keys only: ",
+                  listAllDispatchKeys(), ". Please, make sure the input and",
+                  " the operation are registered with the same dispatch key.");
```
The way this error is worded is still concerning. Remember that this is a highly-user-visible error message. In particular:

1) This error message, with its mention of the concept of "dispatch key", is highly confusing to end users who are simply writing PyTorch Python code. Nowhere in our documentation do we introduce the concept of "dispatch key", which has resulted in numerous issues and forum posts relating to this error message.
2) The error talks about "registration" which, again, is not a concept model authoring users are going to be familiar with. The call to action at the end doesn't make sense from this user viewpoint

Can we please make this error more in-line with something a user would understand, i.e. take into account that viewpoint and not talk about internal implementation details

dzhulgakov replied:
I wonder whether with per-bit boxed fallback in multidispatch (cc @smessmer - I think it's #28494) we can just generate a generic function which gives good user message

And I agree with James - we should think of error messages from user perspective

z-a-f replied:
The concern that @smessmer raised was that the dispatcher should not care what the user is doing, and thus should not throw specialized errors. I tried rewording it so it would give at least a hint to the user on how to fix that.

Here is @smessmer 's comment from #28635:
> I don't think we want this. The core dispatch logic is agnostic of quantization and should stay this way. Quantization depends on dispatch, not the other way round.

jamesr66a replied:
@z-a-f read my comment carefully. The issue I'm bringing up is with the perspective of the error message, not whether it includes quantization or not

z-a-f replied:
@jamesr66a read my comment carefully: my comment is about @smessmer's concern and I am only mirroring it to provide context and show where this PR is coming from.

I agree with your argument, and I would love to change the error message to a more user friendly one. At the same time I agree with Seb about the necessity of keeping information about registered keys in the message. If you have a suggestion on how to accommodate for both -- I will gladly accept it.

smessmer replied:
@dzhulgakov you mean having quantization register a backend level fallback that shows an error message? I had been considering the same thing but it unfortunately doesn't work because backend level fallbacks have a higher priority than operator level fallbacks (i.e. meta ops). So if quantization registers a backend level fallback, meta ops will stop working for quantized tensors.

smessmer replied:
Maybe we can change the wording to say "backend" instead of "dispatch key" and say "available" instead of "registered". cc @jamesr66a

jamesr66a replied:
@smessmer yes that looks good, it covers all off the user perspectives. Though I would prefer if we could make it look something like the existing type checking errors:

jamesr66a reviewed (changes requested):
Requesting changes as we've come to a consensus

Commit:
Changed the error message on "Removing quantization from the dispatcher. Changing the message."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29054

Differential Revision: [D18276881](https://our.internmc.facebook.com/intern/diff/D18276881)

Commit:
Update on "Removing quantization from the dispatcher. Changing the message."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29054

Differential Revision: [D18276881](https://our.internmc.facebook.com/intern/diff/D18276881)

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -291,25 +291,11 @@ class DispatchTable final {
-      TORCH_CHECK(false, "Didn't find kernel to dispatch to for operator '", operator_name_,
-               "'. Tried to look up kernel for dispatch key '", dispatch_key_str,
-               "'. Registered dispatch keys are: ", listAllDispatchKeys());
+      TORCH_CHECK(false, "Could not run '", operator_name_, "' with its input",
```
`with arguments from the ... backend` is probably better than `input` because python developers think in functions not in graph nodes with input connections.

Commit:
Update on "Removing quantization from the dispatcher. Changing the message."


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29054

Differential Revision: [D18276881](https://our.internmc.facebook.com/intern/diff/D18276881)
